### DISCUSSION TITLE
Improve benchmarks

### DISCUSSION THREAD

**Discussion by rafaell-lycan** [honojs/hono#1483] [Ideas]
Hi 👋 

I've been using Hono for a couple of weeks now, and I gotta say it is simple and works!

One of the small things I haven't seen yet is a better benchmark comparison of a non-serverless environment vs. other Node web frameworks such as Express, Hapi, Koa, Restify, and Fastify on [Fly.io](https://fly.io/) or similar.

I came across this proposal after seeing [these results](https://web-frameworks-benchmark.netlify.app/result?l=javascript) and wondered if there's something odd with the tests or is due to Hono's nature of being optimized for serverless.

**Comment by yusukebe** [comment]
Hi @rafaell-lycan 

In the Node environment, Hono is very slow because it uses an Adapter to convert between Web Standard APIs and Node input/output. Therefore, we do not really want to inform them of the results.

But, [the results of this benchmark](https://web-frameworks-benchmark.netlify.app/result?l=javascript) are weird. At least Hono running on Bun is faster than Fastify. Below are the results at hand.

<img width="953" alt="Screenshot 2023-09-20 at 5 49 13" src="https://github.com/honojs/hono/assets/10682/8ec481c0-f975-4445-bf11-a11f0694166a">

Maybe there's a problem with the way it's set up.

**Reply by s-h-a-d-o-w** [reply]
> In the Node environment, Hono is very slow because it uses an Adapter to convert between Web Standard APIs and Node input/output. Therefore, we do not really want to inform them of the results.

That's pretty problematic ethically. People (and their products) should be transparent about their strengths and weaknesses, so that others can make informed decisions. (And it's of course deceptive because people might trust that performance is great in all environments.)
They might choose to use Hono in all environments regardless, if they think that the performance hit is negligible.

**Reply by yusukebe** [reply]
Things have improved since I said that. The [Node.js Adapter](https://github.com/honojs/node-server) is now faster. Therefore, in the following benchmarks, Hono runs on Node.js better than other Node.js frameworks, such as Fastify:

https://web-frameworks-benchmark.netlify.app/result?l=javascript

**Reply by s-h-a-d-o-w** [reply]
Thanks for the update!

**Comment by yusukebe** [comment]
Ah, in that benchmark, Fastify uses cluster to launch multiple processes. This is why it is faster than Hono.

**Reply by yusukebe** [reply]
I just tried it at hand, but it does not run well with pm2.

<img width="1309" alt="Screenshot 2023-09-20 at 6 02 25" src="https://github.com/honojs/hono/assets/10682/6c532036-4f88-45c9-a4ff-a02b912baf5a">

**Reply by yusukebe** [reply]
The solution to this would be to make an adapter for Bun.

**Reply by rafaell-lycan** [reply]
Hey @yusukebe 👋 おはようございます 🌅 

Thanks for dedicating time to this thread.

**QQ:** Does the adapter have any potential drawbacks that might be holding the performance or would be just due to the conversion?

**Why:** I knew that _Hono_ would be faster than _Express_, but I was expecting somewhere near _Restify_ & _Koa_ performance.

This is definitely not the end of the world, and I bet things might get better in the near future by doing small optimizations on the Node adapter, while obviously when using _Bun_ things will probably skyrocket.

I'm buying the whole initiative behind _Hono_ and its simplicity (It reminds me Koa), and I guess that in the mid/long term having a more detailed benchmark (benchmark.md?) would be more explanatory on why using it and where it shines.

ありがとう

**Reply by rafaell-lycan** [reply]
> The solution to this would be to make an adapter for Bun.

Is it worth tho? I'm running a simple server using a `Dockerfile` on Fly.io and it just works.

```dockerfile
FROM oven/bun

WORKDIR /usr/src/app
COPY . .
RUN bun install
EXPOSE 3000

CMD ["bun", "start"]

**Reply by yusukebe** [reply]
Hey, good morning!

It is certainly better to have more benchmarks, but we want to make them as internally oriented as possible. This is because it is very difficult and trouble to publish and maintain benchmarks. Authors of other frameworks may complain about the results. Also, we need to keep them up-to-date. So, we would like to have external benchmarks as much as possible as a reference.

But this is an ideal and just my opinion.

**Reply by yusukebe** [reply]
> Is it worth tho? I'm running a simple server using a `Dockerfile` on Fly.io and it just works.

No problem in that case. But, in that benchmark, Fastify uses multiple processes, because Hono was trying to do that with `pm2`, but was failing. We may need to customize it for Bun to make it work. Need to investigate.

**Comment by stefanos82** [comment]
In my case, the simplest solution was to take advantage of `Bun.spawn()` and `Bun.serve()` to use all CPU cores on my computer:

### cluster.ts

```ts
const numCPUs = navigator.hardwareConcurrency;
const buns = new Array(numCPUs);

for (let i = 0; i < numCPUs; i++) {
    buns[i] = Bun.spawn({
        cmd: ["bun", `${__dirname}/index.ts`],
        stdout: 'inherit',
        stderr: 'inherit',
        stdin: 'inherit',
    });
}

function kill() {
    for (const bun of buns) {
        bun.kill();
    }
}

process.on("SIGINT", kill);
process.on("exit", kill);
```
### index.ts

```ts
import { Hono } from 'hono'

const app = new Hono();
const port = 3000;

app.get('/', (c) => {
  return c.text('Hello Hono!')
})

Bun.serve({
    port: port,
    reusePort: true,
    fetch: app.fetch
});

console.log(`Listening on http://localhost:${port}`);
```

It works exceptionally well, I could say!

![image](https://github.com/user-attachments/assets/0e596d4f-fa68-4773-bbc7-d3c860e4ebd5)
